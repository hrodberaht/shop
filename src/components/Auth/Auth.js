import React, { Component } from 'react';
import ConnectedLoginForm from './LoginForm';

export default class Auth extends Component {
  render() {
    return (
      <div className="login-form">
        <ConnectedLoginForm />
      </div>
    );
  }
}
