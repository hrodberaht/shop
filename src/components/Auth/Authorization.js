import { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { getAuthStatus } from '../../store/auth/selectors';
import { logoutSucces } from '../../store/auth/actionCreator';

export class Authorization extends Component {
  render() {
    const {
      auth, withAuth, withOutAuth, logout, handleLogout,
    } = this.props;
    if (logout) return logout(handleLogout);
    if (auth) return withAuth;
    return withOutAuth;
  }
}

const mapStateToProps = state => ({
  auth: getAuthStatus(state),
});

export default connect(
  mapStateToProps,
  { handleLogout: logoutSucces },
)(Authorization);

Authorization.propTypes = {
  withAuth: PropTypes.element,
  withOutAuth: PropTypes.element,
  logout: PropTypes.func,
  auth: PropTypes.bool.isRequired,
  handleLogout: PropTypes.func,
};

Authorization.defaultProps = {
  logout: null,
  withAuth: null,
  withOutAuth: null,
  handleLogout: null,
};
