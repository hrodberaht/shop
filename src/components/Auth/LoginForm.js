import React, { Component } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import { getAuthError } from '../../store/auth/selectors';
import { login } from '../../store/auth/actionCreator';

export class LoginForm extends Component {
  state = {
    email: '',
    password: '',
  };

  handleChange = (e) => {
    this.setState({ [e.target.id]: e.target.value });
  };

  resetState = () => {
    this.setState({
      email: '',
      password: '',
    });
  };

  handleSubmit = (e) => {
    const { email, password } = this.state;
    const { handleLogin } = this.props;
    e.preventDefault();
    this.resetState();
    handleLogin(email, password);
  };

  render() {
    const { email, password } = this.state;
    const { error } = this.props;
    return (
      <form onSubmit={this.handleSubmit}>
        <label className="form-label" htmlFor="email">
          <p>Email:</p>
          <input
            className="form-input"
            id="email"
            type="text"
            value={email}
            onChange={this.handleChange}
            required
          />
        </label>
        <label className="form-label" htmlFor="password">
          <p>Password:</p>
          <input
            className="form-input"
            id="password"
            type="password"
            value={password}
            onChange={this.handleChange}
            required
          />
        </label>
        <button type="submit">Log In</button>
        <p className="error-text">{error}</p>
      </form>
    );
  }
}

const mapStateToProps = state => ({
  error: getAuthError(state),
});

export default connect(
  mapStateToProps,
  { handleLogin: login },
)(LoginForm);

LoginForm.propTypes = {
  handleLogin: PropTypes.func.isRequired,
  error: PropTypes.string,
};

LoginForm.defaultProps = {
  error: null,
};
