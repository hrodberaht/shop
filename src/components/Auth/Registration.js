import React, { Component } from 'react';
import { Field, reduxForm, SubmissionError } from 'redux-form';
import { Link, withRouter } from 'react-router-dom';
import * as yup from 'yup';
import dataFetcher from '../../shared/dataFetcher';

const schema = yup.object().shape({
  email: yup
    .string()
    .email()
    .required(),
  firstName: yup.string().required(),
  lastName: yup.string().required(),
  password: yup
    .string()
    .required()
    .min(6),
  companyId: yup.string().required(),
});

const asyncValidate = values => schema.validate(values, { abortEarly: false }).catch((err) => {
  const errorsForm = {};
  err.inner.forEach((error) => {
    errorsForm[error.path] = error.type;
  });
  throw errorsForm;
});

export class Registration extends Component {
  state = {
    companies: [],
    loaded: false,
    error: false,
  };

  componentDidMount() {
    this.fetchCompanies();
  }

  fetchCompanies = () => dataFetcher('companies', 'get')
    .then((data) => {
      this.setState({ companies: data, loaded: true });
    })
    .catch(() => {
      this.setState({ error: 'Server not working' });
    });

  addUserToDB = (values) => {
    const errorsForm = {};
    return dataFetcher('registration', 'post', null, values).then((res) => {
      if (res.status === 422) {
        errorsForm.email = res.error;
        return errorsForm;
      }

      return res;
    });
  };

  submit = async (values) => {
    const check = await this.addUserToDB(values);
    if (check.error) {
      return Promise.resolve().then(() => {
        throw new SubmissionError({ email: check.error });
      });
    }
    alert('User added');
    return this.props.history.push('/');
  };

  renderField = ({
    input, label, type, meta: { touched, error },
  }) => (
    <div>
      <input className="form-input" {...input} type={type} placeholder={label} />
      <p className="error-text">{touched && (error && <span>{error}</span>)}</p>
    </div>
  );

  selectOptions = () => (
    <React.Fragment>
      <option />
      {this.state.companies.map(company => (
        <option key={company.id} value={company.id}>
          {company.name}
        </option>
      ))}
    </React.Fragment>
  );

  render() {
    const { handleSubmit, pristine } = this.props;
    if (this.state.error) return <div>{this.state.error}</div>;
    return (
      <div className="registration-form">
        <form onSubmit={handleSubmit(this.submit)}>
          <label className="form-label" htmlFor="email">
            <p>Email:</p>
            <Field name="email" type="text" component={this.renderField} label="Email" />
          </label>
          <label className="form-label" htmlFor="first-name">
            <p>First name:</p>
            <Field name="firstName" type="text" component={this.renderField} label="First name" />
          </label>
          <label className="form-label" htmlFor="last-name">
            <p>Last name:</p>
            <Field name="lastName" type="text" component={this.renderField} label="Last name" />
          </label>
          <label className="form-label" htmlFor="password">
            <p>Password:</p>
            <Field name="password" type="password" component={this.renderField} label="Password" />
          </label>
          <label className="form-label" htmlFor="password">
            <p>Your Company</p>
            <Field className="registration-select" name="companyId" component="select">
              {this.state.loaded && this.selectOptions()}
            </Field>
          </label>
          <button className="btn btn-submit" type="submit" disabled={pristine}>
            Submit
          </button>
          <div className="registration-link">
            <Link to="/">Login</Link>
          </div>
        </form>
      </div>
    );
  }
}

withRouter(Registration);
export default reduxForm({
  asyncValidate,
  form: 'registration',
})(Registration);
