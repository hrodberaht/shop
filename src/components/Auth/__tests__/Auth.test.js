import Auth from '../Auth';

const setup = buildSetup(Auth, {});

describe('<Auth />', () => {
  it('should render without crash', () => {
    const { wrapper } = setup();
    expect(wrapper).toMatchSnapshot();
  });
});
