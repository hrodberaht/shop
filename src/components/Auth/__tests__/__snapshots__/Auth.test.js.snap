// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Auth /> should render without crash 1`] = `
<div
  className="login-form"
>
  <Connect(LoginForm) />
</div>
`;
